types/funds.go

```diff
@@ -0,0 +1,49 @@
+package types
+
+import "math/big"
+
+// IsNonZero returns true if the Holdings structure has any non-zero asset
+func (h Funds) IsNonZero() bool {
+	for _, v := range h {
+		if v.Cmp(big.NewInt(0)) == 1 {
+			return true
+		}
+	}
+	return false
+}
+
+// String returns a bracket-separaged list of assets: {[0x0a,0x01][0x0b,0x01]}
+func (h Funds) String() string {
+	if len(h) == 0 {
+		return "{}"
+	}
+	var s string = "{"
+	for asset, amount := range h {
+		s += "[" + asset.Hex() + "," + amount.Text(64) + "]"
+	}
+	s = s + "}"
+	return s
+}
+
+// todo:
+// ToFunds returns a Funds map from its string representation
+// func ToFunds(s string) Funds {}
+
+// Add sums all assets and returns the result. Does not modify the calling Holdings object
+func (h Funds) Add(a ...Funds) Funds {
+	a = append(a, h)
+	return Sum(a...)
+}
+
+// Sum returns the sum of all input Funds maps
+func Sum(a ...Funds) Funds {
+	sum := Funds{}
+
+	for _, holdings := range a {
+		for asset, amount := range holdings {
+			sum[asset] = sum[asset].Add(sum[asset], amount)
+		}
+	}
+
+	return sum
+}
```

Comment:
They *were* used outside of this file, but a rewrite of some logic in `direct-funing.go` removed it.

Possibly a UK/NA distinction, but `a.Sum(b)` returning the sum of a and b does not read very naturally to me. `Sum` wants (grammatically) an unordered set as input, and having one element differently positioned as the receiver doesn't jive.

Updates to the comments are great. 03807ef

I've done some work on tests (and associated `Equal` function) but will park it for a separate PR (#47).

Reply:
I don't think it is particularly a UK/NA thing. I agree it doesn't jive well. 

* a + b is "a plus b" or "a add b". You've got a binary operator.

* \sum _i c_i is "the sum of" the set of c_i 

but your `a.Add(b)` method is actually `a.Add(b_i)`  -- so it's not a binary operator. If you changed it 
```diff
- func (h Funds) Add(a ...Funds)
+ func (h Funds) Add(a Funds)
```
then it would make more sense to me why it might have a different name.

I think in any case this is not worth us overly worrying about :-)

Reply:
Apologies, I guess you intended to move the conversation to #47. I'll copy my comment there and mark this as resolved.